### ISSUE TITLE
Improvement for collection matching

### ISSUE THREAD

**Issue by almondtools** [joel-costigliola/assertj-core#1256]
#### Summary

Consider a List of complex objects:

* there is more than one element in this list
* the assertion should cover multiple properties, but not the complete state (e.g excluding transient fields like the modification counter in collections)
* each element is different on the asserted properties

Such scenarios are quite easily solved with Hamcrest matchers or AssertJ conditions, yet many developers avoid writing matchers or conditions and go the easy (and fragile) way.

In this [file](https://gist.github.com/almondtools/b884900746cb2028886f50039a30d397) you could find some hacky examples (tests that I do not like) and a proposal how I think collection matching could be more convenient with AssertJ.

I do not know whether it could be implemented properly. Yet I would be interested what you think about it.

#### Example

```java
List<ComplexObject> list = createList();

assertThat(list).satisfyInOrder(element -> {
	assertThat(element.getValue()).isEqualTo("get0");
	assertThat(element.getFixedCounter()).isEqualTo(0);
}, element -> {
	assertThat(element.getValue()).isEqualTo("get1");
	assertThat(element.getFixedCounter()).isEqualTo(0);
}, element -> {
	assertThat(element.getValue()).isEqualTo("get2");
	assertThat(element.getFixedCounter()).isEqualTo(2);
});
```

**Comment by almondtools** [comment]
I noticed some problems with this approach:

* the compiler warnings for unchecked types occur
* error messages are tricky as well
* several overloaded methods will be needed:
  * for Collections
    * for Lists (ordered, index-sequential)
    * for Queues (ordered, no index)
    * for Sets (unordered, unique elements)
    * for SortedSets (ordered, unique elements)
  * for Maps (unordered, keyvalue pairs, unique keys)
  * for arrays

If you think this adds to much noise to the api, you may close this issue.

**Comment by joel-costigliola** [comment]
I agree this would be good additions/improvements but there is way to much work to do in your previous comment, this issue needs to be broken down into several ones with code examples (I had a look at your gist, it is a bit confusing sometimes)  

I would not put it as a high priority issue though (read: contribution wanted) as you can already navigate to the element and use [`satisfies`](http://joel-costigliola.github.io/assertj/core-8/api/org/assertj/core/api/AbstractAssert.html#satisfies-java.util.function.Consumer-) Object assertion. I agree it is not as elegant as  passing a list of lambdas assertions requirements but it gets the job done.

As you mentioned the error message is going to be tricky to build (especially if we report the errors for each elements instead of the first element error).

List assertion kind of supports this with [`satisfies(requirements, index)`](http://joel-costigliola.github.io/assertj/core-8/api/org/assertj/core/api/AbstractListAssert.html#satisfies-java.util.function.Consumer-org.assertj.core.data.Index-) in [3.10.0](http://joel-costigliola.github.io/assertj/assertj-core-news.html#assertj-core-3.10.0-satisfies_for_list).

As a side note, you can write your assertion example like that:
```
assertThat(list).extracting(ComplexObject::getValue, ComplexObject::getFixedCounter)
                .containsExactly(tuple("get0", 0), 
                                 tuple("get1", 0), 
                                 tuple("get2", 2));
```
which I believe is nice if one understands what the tuple are for.

**Comment by almondtools** [comment]
> I would not put it as a high priority issue though (read: contribution wanted) as you can already navigate to the element and use satisfies Object assertion. I agree it is not as elegant as passing a list of lambdas assertions requirements but it gets the job done.

The workaround by navigating to explicity elements is not applicable for:
- lists without stable order (bags)
- sets
- queues
- maps
- ...

Thanks for the hint on tuple extraction. This will technically solve the problem for all of the upper collections/maps. Yet I think it is less readable:

```
the list should contain 
   - an element with value "get0" and counter 0
   - an element with value "get1" and counter 0
   - ...
```

is easier to understand than

```
the list reduced to its attributes value and counter should contain
   - ("get0", 0)
   - ("get1", 0)
   - ...
```

It is not that obvious with 2 properties, but consider scenarios where the elements are 
- characterized by more than 5 properties
- dependent on their static type (each element may provide other properties)
- with wildcards (some properties are not checked on each element)

If I were certain that my proposal above would solve all these problems, I would provide a contribution. I first will try out a solution with conditions.

**Comment by joel-costigliola** [comment]
@almondtools I agree to add a way to run different assertions on different element of a group, let's start with `Iterable` and see how it goes (especially if we can get a clear error message).
